config_utilities/include/config_utilities/factory.h

```diff
@@ -35,184 +35,269 @@
+  bool empty() override { return map.empty(); }
 
-  // Nested modules: base_type -> args -> registered <type_name, tpye>.
-  std::map<std::string, std::map<std::string, std::vector<std::pair<std::string, std::string>>>> modules;
+ private:
+  std::map<std::string, Factory> map;
 };
 
-// Struct to store the factory methods for the creation of modules.
-template <typename FactoryMethod>
-struct ModuleMapBase {
-  using FactoryMethodMap = std::unordered_map<std::string, FactoryMethod>;
-
-  // Add entries to the map with verbose warnings.
-  static bool addEntry(const std::string& type, const FactoryMethod& method, const std::string& type_info) {
-    FactoryMethodMap& map = instance().map;
-    if (map.find(type) != map.end()) {
-      if (!type_info.empty()) {
-        Logger::logError("Cannot register already existent type '" + type + "' for " + type_info + ".");
+class ModuleRegistry {
+ public:
+  using LockCallback = std::function<void(const ModuleInfo&, std::string, std::string)>;
+  using CreateCallback = std::function<void(const ModuleInfo&, std::string, void*)>;
+
+  template <typename BaseT, typename DerivedT, typename... Args>
+  static bool addModule(const std::string& type, FactoryMethod<BaseT, Args...> method, bool skip_first_arg = false) {
+    const auto key = ModuleInfo::fromTypes<BaseT, Args...>(skip_first_arg);
+    using Factory = FactoryMethod<BaseT, Args...>;
+    if (locked()) {
+      if (hasModule(key, type)) {
+        return false;
+      }
+
+      const auto& callback = instance().lock_callback_;
+      if (callback) {
+        callback(key, type, typeName<DerivedT>());
       }
+    }
+
+    auto& modules = instance().modules;
+    auto iter = modules.find(key);
+    if (iter == modules.end()) {
+      iter = modules.emplace(key, std::make_unique<FactoryMap<Factory>>()).first;
+    }
+
+    auto derived = dynamic_cast<FactoryMap<Factory>*>(iter->second.get());
+    if (!derived) {
+      Logger::logFatal("Invalid module map for type '" + type + "' and info " + key.typeInfo());
+      return false;
+    }
+
+    if (!derived->addEntry(type, method)) {
+      Logger::logError("Cannot register already existent type '" + type + "' for " + key.typeInfo() + ".");
       return false;
     }
 
-    map.insert(std::make_pair(type, method));
+    instance().type_registry[key][type] = typeName<DerivedT>();
     return true;
   }
 
-  // Check if a requested type is valid with verbose warnings.
-  static bool hasEntry(const std::string& type, const std::string& type_info, const std::string& registration_info) {
-    FactoryMethodMap& map = instance().map;
-    if (map.empty()) {
+  template <typename BaseT, typename... Args>
+  static FactoryMethod<BaseT, Args...> getModule(const std::string& type,
+                                                 const std::string& registration_info,
+                                                 bool skip_first_arg = false) {
+    using Factory = FactoryMethod<BaseT, Args...>;
```

Comment:
Nitpick: factories and factory methods are technically not the same thing (and typically similar but not identical patterns). I don't think people will be too confused here though.